In [1]:
import pandas as pd

# A. Introduction to Pandas DataFrame and Transformation (45pts)

## Beginner

### 1. Membuat DataFrame Manual (5pts)

Buatlah dataframe secara manual seperti gambar dibawah ini.

**Expected Result:**

![Expected Result](assets/homework_exercise_1_result.png)

In [2]:
## Your code below here
import pandas as pd
dataframe = pd.DataFrame({"name" : ["fiqry", "iqbal", "monica", "rama", "johan"],
                          "age"  : [23,21,22,24,26],
                          "phone_number": ["+62813123414","+6287842464","+62813125554","+6287834464","+62813113414"],
                          "ielts_score" : [6.5, None, 7.5, 6.5, 8.0]
                    })
dataframe

name  age  phone_number  ielts_score
0   fiqry   23  +62813123414          6.5
1   iqbal   21   +6287842464          NaN
2  monica   22  +62813125554          7.5
3    rama   24   +6287834464          6.5
4   johan   26  +62813113414          8.0

### 2. Analisa Sederhana DataFrame (10pts)

Lakukan import `application_processed.csv` data kedalam notebook (5pts) dan carilah beberapa informasi pada dataset `application_processed` (5pts) dan jawablah beberapa pertanyaan sebagai berikut:

a. Column apa saja yang memiliki nilai NaN / None?
b. Berapa rata-rata pada column `amount_income_total`?
c. Berapa median pada column `amount_income_total`?

**Hint**:
- Menggunakan operasi dasar DataFrame pada topik Introduction Pandas DataFrame and Transformation


In [3]:
## Your code below here
df = pd.read_csv("application_processed.csv")
df.head()

id  gender  flag_own_car  flag_own_property  total_children  \
0  5548815  Female             1                  1               1   
1  5692601  Female             1                  1               0   
2  5625007  Female             1                  0               0   
3  6529653    Male             0                  1               0   
4  6209935    Male             0                  1               1   

   amount_income_total           income_type                 education_type  \
0             225000.0         State servant  Secondary / secondary special   
1             180000.0               Working  Secondary / secondary special   
2             157500.0             Pensioner  Secondary / secondary special   
3             202500.0               Working  Secondary / secondary special   
4             157500.0  Commercial associate  Secondary / secondary special   

  family_status       housing_type  days_birth  days_employed  \
0       Married  House / apartment      -12944          -6127   
1       Married  House / apartment      -12484          -1874   
2         Widow  House / apartment      -21656         365243   
3       Married  House / apartment      -10050          -2117   
4       Married  House / apartment      -10371          -1755   

   flag_mobile_phone  flag_work_phone  flag_phone  flag_email occupation_type  \
0                  1                0           0           0        Laborers   
1                  1                0           0           0             NaN   
2                  1                0           0           0             NaN   
3                  1                0           0           0         Drivers   
4                  1                0           0           0         Drivers   

   total_family_members  
0                   3.0  
1                   2.0  
2                   1.0  
3                   2.0  
4                   3.0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131567 entries, 0 to 131566
Data columns (total 18 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   id                    131567 non-null  int64  
 1   gender                131567 non-null  object 
 2   flag_own_car          131567 non-null  int64  
 3   flag_own_property     131567 non-null  int64  
 4   total_children        131567 non-null  int64  
 5   amount_income_total   131567 non-null  float64
 6   income_type           131567 non-null  object 
 7   education_type        131567 non-null  object 
 8   family_status         131567 non-null  object 
 9   housing_type          131567 non-null  object 
 10  days_birth            131567 non-null  int64  
 11  days_employed         131567 non-null  int64  
 12  flag_mobile_phone     131567 non-null  int64  
 13  flag_work_phone       131567 non-null  int64  
 14  flag_phone            131567 non-null  int64  
 15  

In [5]:
missing_value = df.isna().sum()
print(missing_value)

id                          0
gender                      0
flag_own_car                0
flag_own_property           0
total_children              0
amount_income_total         0
income_type                 0
education_type              0
family_status               0
housing_type                0
days_birth                  0
days_employed               0
flag_mobile_phone           0
flag_work_phone             0
flag_phone                  0
flag_email                  0
occupation_type         40095
total_family_members        0
dtype: int64


In [6]:
df.describe()

id   flag_own_car  flag_own_property  total_children  \
count  1.315670e+05  131567.000000      131567.000000   131567.000000   
mean   6.021734e+06       0.374547           0.691496        0.427653   
std    5.720755e+05       0.484008           0.461878        0.724228   
min    5.008810e+06       0.000000           0.000000        0.000000   
25%    5.603155e+06       0.000000           0.000000        0.000000   
50%    6.048173e+06       0.000000           1.000000        0.000000   
75%    6.459922e+06       1.000000           1.000000        1.000000   
max    7.999660e+06       1.000000           1.000000       19.000000   

       amount_income_total     days_birth  days_employed  flag_mobile_phone  \
count         1.315670e+05  131567.000000  131567.000000           131567.0   
mean          1.880406e+05  -15992.029323   60662.604665                1.0   
std           1.150469e+05    4183.265031  138861.427009                0.0   
min           2.700000e+04  -25201.000000  -17531.000000                1.0   
25%           1.215000e+05  -19472.000000   -3109.000000                1.0   
50%           1.620000e+05  -15617.000000   -1467.000000                1.0   
75%           2.250000e+05  -12496.000000    -371.000000                1.0   
max           6.750000e+06   -7678.000000  365243.000000                1.0   

       flag_work_phone     flag_phone     flag_email  total_family_members  
count    131567.000000  131567.000000  131567.000000         131567.000000  
mean          0.207453       0.290308       0.107299              2.193954  
std           0.405484       0.453906       0.309494              0.896506  
min           0.000000       0.000000       0.000000              1.000000  
25%           0.000000       0.000000       0.000000              2.000000  
50%           0.000000       0.000000       0.000000              2.000000  
75%           0.000000       1.000000       0.000000              3.000000  
max           1.000000       1.000000       1.000000             20.000000

In [7]:
rata_rata_amount_income_total = df.agg({"amount_income_total": ["mean", "median"]})
rata_rata_amount_income_total

amount_income_total
mean          188040.576437
median        162000.000000

Jawab :

a. Column apa saja yang memiliki nilai NaN / None? 
Dengan menggunakan fungsi info() dapat diketahui bahwa pada dataset application_processed terdapat 1 column yang memiliki missing value/NaN/None yaitu occupation_type

b. Berapa rata-rata pada column amount_income_total?
Dengan menggunakan fungsi describe() dapat diketahui bahwa pada colum amount_income_total memiliki rata-rata sebesar 1.880406e+05 atau 188040,6

c. Berapa median pada column amount_income_total?
Dengan menggunakan fungsi describe() dapat diketahui bahwa pada colum amount_income_total memiliki median sebesar 1.620000e+05 atau 162000

## Intermediate

### 3. Mengolah Dataset (30pts)

Dengan menggunakan dataset `application_processed.csv` buatlah dataframe baru dengan ketentuan:
1. Ambil semua users yang memiliki `email`. (ditandai dengan nilai pada column `flag_email` = 1) dan tidak memilik data `duplicates`. (5pts)
2. Buatlah column baru bernama `has_car_and_property` dimana nilainya merupakan range dari `0-1`. `1` apabila `flag_own_car` bernilai `1` dan `flag_own_property` bernilai `1`. `0` apabila terdapat atau semua nilai pada column `flag_own_car` dan `flag_own_property` terdapat angka 0. (5pts)
3. Ambil column yang diperlukan yaitu `id`, `gender`, `has_car_and_property`, `income_type`, `education_type`, `family_status`, `housing_type`, `occupation_type`, `amount_income_total`. (5pts)
4. Lakukan filter dengan ketentuan sebagai berikut: (5pts)
    - Memiliki `car` dan `property`
    - Tidak ada nilai null pada column `occupation_type`
    - Memiliki value pada column `occupation_type` dengan pola `staff`.
5. Sudah dilakukan sorting berdasarkan `income_type` dari `A-Z` dan `amount_income_total` dari `besar ke kecil`. (5pts)
6. Simpan kedalam bentuk `.csv` dengan nama `homework_3_result.csv`. (5pts)

**Expected Result:**

![Expected Result](assets/homework_exercise_3_result.png)

In [8]:
## Your code below here
df = pd.read_csv("application_processed.csv")

In [9]:
df1 = df
df1 = df1[df1["flag_email"]==1]
# mengecek apakah ada duplicate
df1.duplicated().any()
# mengecek baris mana yang ada duplicate
df1[df1.duplicated(keep = False) == True]
# menghapus data duplicate 
df1 = df1.drop_duplicates()
df1

id  gender  flag_own_car  flag_own_property  total_children  \
5       5541414  Female             0                  1               0   
8       5169560  Female             1                  1               0   
22      5884982  Female             0                  1               0   
23      6297619    Male             1                  1               0   
28      6006818  Female             0                  1               0   
...         ...     ...           ...                ...             ...   
131511  5345596    Male             1                  1               1   
131514  5680139  Female             0                  0               1   
131518  5708288  Female             0                  1               2   
131537  5352502  Female             1                  1               0   
131555  6234604  Female             1                  1               1   

        amount_income_total           income_type  \
5                  211500.0               Working   
8                   90000.0               Working   
22                 337500.0  Commercial associate   
23                 360000.0  Commercial associate   
28                 247500.0             Pensioner   
...                     ...                   ...   
131511             112500.0               Working   
131514              67500.0               Working   
131518             135000.0               Working   
131537             103500.0  Commercial associate   
131555             171000.0               Working   

                       education_type         family_status  \
5       Secondary / secondary special        Civil marriage   
8       Secondary / secondary special               Married   
22      Secondary / secondary special               Married   
23                   Higher education  Single / not married   
28                   Higher education               Married   
...                               ...                   ...   
131511               Higher education               Married   
131514  Secondary / secondary special               Married   
131518               Higher education               Married   
131537               Higher education        Civil marriage   
131555  Secondary / secondary special               Married   

               housing_type  days_birth  days_employed  flag_mobile_phone  \
5         House / apartment      -18171          -1671                  1   
8         House / apartment      -16023          -3532                  1   
22      Municipal apartment      -15491           -703                  1   
23         Rented apartment      -15796          -1625                  1   
28        House / apartment      -23032         365243                  1   
...                     ...         ...            ...                ...   
131511    House / apartment      -17318           -246                  1   
131514    House / apartment      -15140          -2859                  1   
131518         With parents      -13296          -2178                  1   
131537    House / apartment      -11772           -113                  1   
131555     Office apartment      -13908           -993                  1   

        flag_work_phone  flag_phone  flag_email        occupation_type  \
5                     0           1           1                    NaN   
8                     0           0           1            Sales staff   
22                    0           1           1               Managers   
23                    0           0           1  High skill tech staff   
28                    0           0           1                    NaN   
...                 ...         ...         ...                    ...   
131511                0           0           1               Managers   
131514                0           0           1          Cooking staff   
131518                0           0           1  High skill tech staff   
131537                0           0      

In [10]:
df2 = df1
df2["has_car_and_property"] = df2["flag_own_car"] * df2["flag_own_property"]
df2

id  gender  flag_own_car  flag_own_property  total_children  \
5       5541414  Female             0                  1               0   
8       5169560  Female             1                  1               0   
22      5884982  Female             0                  1               0   
23      6297619    Male             1                  1               0   
28      6006818  Female             0                  1               0   
...         ...     ...           ...                ...             ...   
131511  5345596    Male             1                  1               1   
131514  5680139  Female             0                  0               1   
131518  5708288  Female             0                  1               2   
131537  5352502  Female             1                  1               0   
131555  6234604  Female             1                  1               1   

        amount_income_total           income_type  \
5                  211500.0               Working   
8                   90000.0               Working   
22                 337500.0  Commercial associate   
23                 360000.0  Commercial associate   
28                 247500.0             Pensioner   
...                     ...                   ...   
131511             112500.0               Working   
131514              67500.0               Working   
131518             135000.0               Working   
131537             103500.0  Commercial associate   
131555             171000.0               Working   

                       education_type         family_status  \
5       Secondary / secondary special        Civil marriage   
8       Secondary / secondary special               Married   
22      Secondary / secondary special               Married   
23                   Higher education  Single / not married   
28                   Higher education               Married   
...                               ...                   ...   
131511               Higher education               Married   
131514  Secondary / secondary special               Married   
131518               Higher education               Married   
131537               Higher education        Civil marriage   
131555  Secondary / secondary special               Married   

               housing_type  days_birth  days_employed  flag_mobile_phone  \
5         House / apartment      -18171          -1671                  1   
8         House / apartment      -16023          -3532                  1   
22      Municipal apartment      -15491           -703                  1   
23         Rented apartment      -15796          -1625                  1   
28        House / apartment      -23032         365243                  1   
...                     ...         ...            ...                ...   
131511    House / apartment      -17318           -246                  1   
131514    House / apartment      -15140          -2859                  1   
131518         With parents      -13296          -2178                  1   
131537    House / apartment      -11772           -113                  1   
131555     Office apartment      -13908           -993                  1   

        flag_work_phone  flag_phone  flag_email        occupation_type  \
5                     0           1           1                    NaN   
8                     0           0           1            Sales staff   
22                    0           1           1               Managers   
23                    0           0           1  High skill tech staff   
28                    0           0           1                    NaN   
...                 ...         ...         ...                    ...   
131511                0           0           1               Managers   
131514                0           0           1          Cooking staff   
131518                0           0           1  High skill tech staff   
131537                0           0      

In [12]:
df3 = df2
df3 = df3.drop(["flag_own_car","flag_own_property","total_children","days_birth","days_employed",
              "flag_mobile_phone","flag_work_phone","flag_phone","flag_email","total_family_members"], axis=1)
df3

id  gender  amount_income_total           income_type  \
5       5541414  Female             211500.0               Working   
8       5169560  Female              90000.0               Working   
22      5884982  Female             337500.0  Commercial associate   
23      6297619    Male             360000.0  Commercial associate   
28      6006818  Female             247500.0             Pensioner   
...         ...     ...                  ...                   ...   
131511  5345596    Male             112500.0               Working   
131514  5680139  Female              67500.0               Working   
131518  5708288  Female             135000.0               Working   
131537  5352502  Female             103500.0  Commercial associate   
131555  6234604  Female             171000.0               Working   

                       education_type         family_status  \
5       Secondary / secondary special        Civil marriage   
8       Secondary / secondary special               Married   
22      Secondary / secondary special               Married   
23                   Higher education  Single / not married   
28                   Higher education               Married   
...                               ...                   ...   
131511               Higher education               Married   
131514  Secondary / secondary special               Married   
131518               Higher education               Married   
131537               Higher education        Civil marriage   
131555  Secondary / secondary special               Married   

               housing_type        occupation_type  has_car_and_property  
5         House / apartment                    NaN                     0  
8         House / apartment            Sales staff                     1  
22      Municipal apartment               Managers                     0  
23         Rented apartment  High skill tech staff                     1  
28        House / apartment                    NaN                     0  
...                     ...                    ...                   ...  
131511    House / apartment               Managers                     1  
131514    House / apartment          Cooking staff                     0  
131518         With parents  High skill tech staff                     0  
131537    House / apartment            Accountants                     1  
131555     Office apartment            Sales staff                     1  

[12235 rows x 9 columns]

In [13]:
df4 = df3
df4 = df4[df4["has_car_and_property"]==1]
df4 = df4[df4["occupation_type"].notnull()]
df4 = df4[df4["occupation_type"].str.contains("staff")]
df4

id  gender  amount_income_total           income_type  \
8       5169560  Female              90000.0               Working   
23      6297619    Male             360000.0  Commercial associate   
169     5548993  Female             405000.0  Commercial associate   
276     5344333  Female             135000.0               Working   
353     5499766  Female             225000.0               Working   
...         ...     ...                  ...                   ...   
131412  6046953    Male             450000.0         State servant   
131439  6484818    Male             315000.0               Working   
131470  5883006  Female             157500.0         State servant   
131481  6728550  Female             157500.0               Working   
131555  6234604  Female             171000.0               Working   

                       education_type         family_status  \
8       Secondary / secondary special               Married   
23                   Higher education  Single / not married   
169                  Higher education               Married   
276     Secondary / secondary special               Married   
353     Secondary / secondary special        Civil marriage   
...                               ...                   ...   
131412               Higher education  Single / not married   
131439               Higher education               Married   
131470               Higher education  Single / not married   
131481              Incomplete higher        Civil marriage   
131555  Secondary / secondary special               Married   

             housing_type        occupation_type  has_car_and_property  
8       House / apartment            Sales staff                     1  
23       Rented apartment  High skill tech staff                     1  
169     House / apartment             Core staff                     1  
276     House / apartment         Medicine staff                     1  
353     House / apartment             Core staff                     1  
...                   ...                    ...                   ...  
131412  House / apartment             Core staff                     1  
131439  House / apartment            Sales staff                     1  
131470  House / apartment         Medicine staff                     1  
131481  House / apartment             Core staff                     1  
131555   Office apartment            Sales staff                     1  

[1258 rows x 9 columns]

In [14]:
df5 = df4
df5 = df5.sort_values(["income_type","amount_income_total"], ascending = [True, False])
df5

id  gender  amount_income_total           income_type  \
52421   6604643    Male            1350000.0  Commercial associate   
59763   6604646    Male            1350000.0  Commercial associate   
64362   6604645    Male            1350000.0  Commercial associate   
24436   6423956  Female             837000.0  Commercial associate   
53672   6423944  Female             837000.0  Commercial associate   
...         ...     ...                  ...                   ...   
52545   6659184  Female              67500.0               Working   
52559   5021814  Female              67500.0               Working   
80932   5193900  Female              67500.0               Working   
117940  6659179  Female              67500.0               Working   
73293   5936326    Male              40500.0               Working   

                       education_type         family_status  \
52421   Secondary / secondary special  Single / not married   
59763   Secondary / secondary special  Single / not married   
64362   Secondary / secondary special  Single / not married   
24436                Higher education               Married   
53672                Higher education               Married   
...                               ...                   ...   
52545   Secondary / secondary special               Married   
52559   Secondary / secondary special               Married   
80932                Higher education               Married   
117940  Secondary / secondary special               Married   
73293   Secondary / secondary special               Married   

             housing_type occupation_type  has_car_and_property  
52421   House / apartment      Core staff                     1  
59763   House / apartment      Core staff                     1  
64362   House / apartment      Core staff                     1  
24436   House / apartment      Core staff                     1  
53672   House / apartment      Core staff                     1  
...                   ...             ...                   ...  
52545   House / apartment      Core staff                     1  
52559   House / apartment   Cooking staff                     1  
80932   House / apartment     Sales staff                     1  
117940  House / apartment      Core staff                     1  
73293   House / apartment  Security staff                     1  

[1258 rows x 9 columns]

In [15]:
df6 = df5
df6 = df6.to_csv("homework_3_result.")

# B.DataFrame Aggregation (25pts)

## Intermediate

### 4. Group By and Aggregate (10pts)

Dengan menggunakan data `titanic.csv`, buatlah dataframe untuk melihat informasi `max` dan `min` pada column `Age` dan juga informasi `mean` dan `median` pada column `Fare` pada masing-masing PassengerClass atau `Pclass`.

**Expected Result:**

![Expected Result](assets/homework_exercise_4_result.png)

In [16]:
## Your code below here
dataframetitanic = pd.read_csv("titanic.csv")
dataframetitanic.groupby("Pclass").agg({"Age" : ["max","min"],
                                       "Fare" : ["mean","median"]})

Age             Fare         
         max   min       mean   median
Pclass                                
1       80.0  0.92  84.154687  60.2875
2       70.0  0.67  20.662183  14.2500
3       74.0  0.42  13.675550   8.0500

### 5. Memanfaatkan Pivot Tabel (15pts)

Buatlah dataframe baru menggunakan fungsi pivot table untuk melihat `Embarked` apa saja yang `average_female_ticket_price` lebih besar daripada `average_male_ticket_price`. Informasi `average_female_ticket_price` dan `average_male_ticket_price` didapatkan dari rata-rata pada column `Fare`.

**Hint:**
1. menggunakan pivot table
2. dapat dilihat pada topik `DataFrame Aggregation` bagian `Pivot Table | Reshape Columns and Rows`

**Expected Result:**

![Expected Result](assets/homework_exercise_5_result.png)

In [18]:
## Your code below here
dataframetitanic = pd.read_csv("titanic.csv")
dataframetitanic.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [22]:
dataframetitanic= pd.pivot_table(dataframetitanic,
                                 index = ["Embarked"],
                                 columns = ["Sex"],
                                 values = ["Fare"],
                                 aggfunc = {"Fare" : ["mean"]}
)
dataframetitanic = dataframetitanic.reset_index()
dataframetitanic.columns = ["Embarked", "avg_female_ticket_price","avg_male_ticket_price"]
dataframetitanic[dataframetitanic["avg_female_ticket_price"] > dataframetitanic["avg_male_ticket_price"]]

Embarked  avg_female_ticket_price  avg_male_ticket_price
0        C                75.169805              48.262109
2        S                38.740929              21.711996

# C. DataFrame Combination (30pts)

## 6. Melakukan Append pada DataFrame (5pts)

Buatlah DataFrame seperti pada soal `no.1`. Kemudian tambahkan data tersebut dengan data users yang baru seperti gambar dibawah berikut.

**New Users Data:**

![Expected Result](assets/homework_exercise_6_new.png)

**Expected Result:**

![Expected Result](assets/homework_exercise_6_result.png)

In [23]:
# No 6 
dataframe = pd.DataFrame({"name" : ["fiqry", "iqbal", "monica", "rama", "johan"],
                          "age"  : [23,21,22,24,26],
                          "phone_number": ["+62813123414","+6287842464","+62813125554","+6287834464","+62813113414"],
                          "ielts_score" : [6.5, None, 7.5, 6.5, 8.0]})
new_users_data = pd.DataFrame({"name" : ["ali","adit"],
                               "age"  : [37,32],
                               "phone_number" : [None, "+62152155"],
                               "ielts_score"  : [5.5,6.0]})
dataframe.append(new_users_data)

name  age  phone_number  ielts_score
0   fiqry   23  +62813123414          6.5
1   iqbal   21   +6287842464          NaN
2  monica   22  +62813125554          7.5
3    rama   24   +6287834464          6.5
4   johan   26  +62813113414          8.0
0     ali   37          None          5.5
1    adit   32     +62152155          6.0

### 7. Merge DataFrame (25pts)

Buatlah 2 dataframe mengikuti petunjuk berikut:

df_1 :  Filter dataset 01 telecom_revenue.csv dengan multiple kondisi (5pts)
- MonthlyRevenue > 10 
- Occupation terdiri dari Professional, Student, and Crafts

df_2 :  Filter dataset 02 telecom_usage.csv dengan multiple kondisi (5pt)
- nilai UnansweredCalls > BlockedCalls

Jawablah pertanyaan berikut ini:
1. Hitunglah total CustomerID dan rata-rata DroppedCalls untuk masing-masing occupation. (10 Points)
2. Pada occupation apa, rata-rata DroppedCalls paling besar? (5 Points)

**Expected Result:**

![Expected Result](assets/homework_exercise_7_result.png)

In [24]:
# No 7
df_1 = pd.read_csv("01 telecom_revenue.csv")
df_1 = df_1[df_1["MonthlyRevenue"] > 10]
df_1 = df_1[df_1["Occupation"].isin(["Professional", "Student", "Crafts"])]
df_1
df_2 = pd.read_csv("02 telecom_usage.csv")
df_2 = df_2[df_2["UnansweredCalls"] > df_2["BlockedCalls"]]
df_2
data_merge = df_2.merge(df_1, left_on = "CustomerID", right_on = "CustomerID", how = "inner")
data_merge

CustomerID Churn  MonthlyRevenue  MonthlyMinutes  TotalRecurringCharge  \
12       3397170    No           83.51           290.0                  93.0   
25       3274822    No           30.83           273.0                  40.0   
26       3394258    No           52.72          1501.0                  62.0   
29       3056178   Yes           54.09           412.0                  40.0   
35       3065718    No           43.06           110.0                  35.0   
...          ...   ...             ...             ...                   ...   
9964     3015822    No           46.10           203.0                  38.0   
9968     3323878    No           44.26           419.0                  45.0   
9970     3050290   Yes           51.06           192.0                  30.0   
9981     3119258    No           39.49           352.0                  45.0   
9994     3125862    No           22.50            18.0                  17.0   

      OverageMinutes  RoamingCalls    Occupation  
12               0.0           0.0  Professional  
25               0.0           0.8  Professional  
26               0.0           0.0        Crafts  
29              46.0           0.0  Professional  
35               0.0           3.3        Crafts  
...              ...           ...           ...  
9964            30.0           0.0        Crafts  
9968             2.0           0.0  Professional  
9970            74.0           0.0  Professional  
9981             0.0           1.7        Crafts  
9994             2.0           0.0  Professional  

[2023 rows x 8 columns]

In [25]:
df_2 = pd.read_csv("02 telecom_usage.csv")
df_2 = df_2[df_2["UnansweredCalls"] > df_2["BlockedCalls"]]
df_2

CustomerID  DroppedCalls  BlockedCalls  UnansweredCalls  \
0        3389874           0.0           1.0              7.7   
1        3282042          15.3           1.7            152.0   
2        3054454           8.7           0.3             30.3   
3        3091842           6.0           1.7             19.3   
4        3382382           0.7           0.7              5.0   
...          ...           ...           ...              ...   
4994     3148902           0.3           0.0              4.3   
4995     3302462           6.7           0.0              4.7   
4997     3066482           0.0           3.0              3.3   
4998     3059734           0.3           0.7              2.0   
4999     3224474           6.7           1.0             32.7   

      CustomerCareCalls  
0                   0.7  
1                   5.7  
2                   0.0  
3                   0.0  
4                   0.0  
...                 ...  
4994                0.0  
4995                0.0  
4997                0.0  
4998                0.0  
4999                0.0  

[4241 rows x 5 columns]

In [29]:
data_merge = df_2.merge(df_1, left_on = "CustomerID", right_on = "CustomerID", how = "inner")
data_merge

CustomerID  DroppedCalls  BlockedCalls  UnansweredCalls  \
0       3252294          15.7          16.0            135.0   
1       3286650           7.0           0.3             20.3   
2       3063250          27.0           0.0             11.3   
3       3326638           8.7           0.3             29.0   
4       3369506           1.3           0.0             33.3   
..          ...           ...           ...              ...   
152     3245094           1.7           0.3              3.7   
153     3033890           1.3           4.3             24.3   
154     3236314           9.3           0.7             35.3   
155     3025750           3.3           1.0             30.7   
156     3208570           3.0           1.3             12.3   

     CustomerCareCalls Churn  MonthlyRevenue  MonthlyMinutes  \
0                  0.0   Yes           55.04           860.0   
1                  1.7    No           59.97           367.0   
2                  0.0    No          250.14          1435.0   
3                  2.0    No           54.18           734.0   
4                  0.0   Yes           38.74           228.0   
..                 ...   ...             ...             ...   
152                0.0    No           11.28            81.0   
153                3.3    No           19.19           278.0   
154                0.3    No           36.89           544.0   
155                0.7   Yes           35.52           216.0   
156                0.0   Yes           39.43           309.0   

     TotalRecurringCharge  OverageMinutes  RoamingCalls    Occupation  
0                    60.0            41.0           0.0  Professional  
1                    70.0             0.0           3.0        Crafts  
2                   100.0           405.0           0.0  Professional  
3                    70.0             0.0           0.0        Crafts  
4                    30.0             0.0           0.0  Professional  
..                    ...             ...           ...           ...  
152                  10.0             0.0           0.3  Professional  
153                  15.0            10.0           0.0  Professional  
154                  30.0            18.0           0.0  Professional  
155                  35.0             9.0           0.0  Professional  
156                  10.0           101.0           1.1  Professional  

[157 rows x 12 columns]

In [30]:
data_group = data_merge.groupby("Occupation").agg({"CustomerID" : ["nunique"],
                                       "DroppedCalls" : ["mean"]})
data_group

CustomerID DroppedCalls
                nunique         mean
Occupation                          
Crafts               24     6.191667
Professional        128     5.532813
Student               5     8.220000

In [31]:
data_group["CustomerID"]["nunique"].sum()

157

1. Hitunglah total CustomerID dan rata-rata DroppedCalls untuk masing-masing occupation. 

Jawab :
Total CustomerID adalah 157.
Rata-rata DroppedCalls untuk Occupation Craft adalah 6,19.
Rata-rata DroppedCalls untuk Occupation Professional adalah 5,53.
Rata-rata DroppedCalls untuk Occupation Student adalah 8,22.

2. Pada occupation apa, rata-rata DroppedCalls paling besar?

Jawab :
rata-rata DroppedCalls paling besar adalah pada Occupation Student